projects/concurrency/README.md

```diff
@@ -0,0 +1,308 @@
+
+## Projects
+
+### Single-variable concurrency
+
+Let's look at the example buggy code from the Mutexes chapter of golangbot:
+
+```go
+package main
+
+import (
+	"fmt"
+	"sync"
+)
+
+var x = 0
+
+func increment(wg *sync.WaitGroup) {
+	x = x + 1
+	wg.Done()
+}
+
+func main() {
+	var w sync.WaitGroup
+	for i := 0; i < 1000; i++ {
+		w.Add(1)
+		go increment(&w)
+	}
+	w.Wait()
+	fmt.Println("final value of x", x)
+}
+```
+
+In an ideal world, this program would always output 1000.
+
+In reality, if you run it 5 times, you will probably get 5 different answers.
+
+The root cause is that multiple threads are handling the variable `x`, and nothing is making sure they don't get in each other's way.
+
+The statement `x = x + 1` does three operations:
+1. Look up the value at the memory location of `x`
+2. Compute `(that looked-up value) + 1`
+3. Assign `(that computed value)` to the memory location of `x`
+
+A problem here is that other operations from other threads may happen between these operations.
+
+Two threads may both read `0` from step 1, compute `1` from step 2, and try to write `1` in step 2 (they did the same work).
+
+Or one thread may read `0`, all other 999 threads may do their work as we hope (incrementing `x` one at a time in turn to `999`), then the first thread may do its step 3 and write `1` back to `x`, undoing the work of the other threads.
+
+There are three families of solutions to this problem:
+1. We can make sure no other thread is allowed to start doing these operations until one thread has completed all three. golangbot showed you how to do this with a `Mutex`.
+2. We can make sure all the operations actually happen from one goroutine. golangbot showed you how to do this with a `channel`.
+3. We can ask the CPU to do one operation which will achieve all three effects without allowing any gaps between them. golangbot hasn't shown you this yet. Let's explore it now.
+
+#### Atomics
+
+The problem we've identified is that we may end up using a stale version of the `x` variable. After we look it up, it may change before we write back our new value.
+
+Atomic operations allow us to avoid this situation.
+
+"Atomic" means "as one unit which can't be broken down". An atomic operation can't be broken down into smaller operations.
+
+Go exposes atomic operations in [the sync/atomic package](https://pkg.go.dev/sync/atomic). [The `sync/atomic.Int32` type](https://pkg.go.dev/sync/atomic#Int32) exposes [an `Add` function](https://pkg.go.dev/sync/atomic#Int32.Add) which does the same as `+=` does to `int`, but looks a little different. Let's compare:
+
+```go
+var x int
+x += 1
+```
+
+```go
+var x atomic.Int32
+newX := x.Add(1)
+```
+
+Both of these do the same thing. They start `x` with value `0`, they add one to it, and at the end `x` will have the value `1`.
+
+The atomic version, though, does so as one operation. It's impossible for another thread to do something in the middle of the steps. The non-atomic version doesn't.
```

Comment:
It could be nice as an aside to link off to https://en.wikipedia.org/wiki/Compare-and-swap
(or another source if you prefer) and to explain that CAS machine instructions are used to implement not only atomics but also mutexes and so on.